subprojects/core-api/src/main/java/org/gradle/api/artifacts/capability/FeatureCapabilitySelector.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts.capability;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A capability selector that dynamically derives the capability
+ * to request based on the module coordinates of the component
+ * being selected and the name of the feature of the component to depend on.
+ * <p>
+ * For example, if this a feature capability selector with feature name 'test-fixtures'
+ * is attached to a component selector that resolves to a component with module coordinates
+ * 'group:name:version', the resolved capability will be 'group:name-test-fixtures:version'.
+ *
+ * @since 8.11
+ */
+@Incubating
+public interface FeatureCapabilitySelector extends CapabilitySelector {
```

Comment:
This is a public API. We generally do not expose _any_ classes as public API -- though there are some legacy cases where we do not follow this pattern.

The main reason is we don't want users instantiating these capabilities themselves. If this was internal code I would definitely agree the interface + implementation is over-engineering but since this is public API I definitely think we want an interface here